pylops/optimization/solver.py

```diff
@@ -2,10 +2,202 @@
+            msg = f"{self.iiter:6g}        {self.x[0]:11.4e}        {self.cost[self.iiter]:11.4e}"
+        else:
+            msg = f"{self.iiter:6g}    {np.real(self.x[0]):4.1e} + {np.imag(self.x[0]):4.1e}j    {self.cost[self.iiter]:11.4e}"
+        print(msg)
+
+    def _print_finalize(self):
+        print(
+            f"\nIterations = {self.iiter}        Total time (s) = {time.time() - self.tstart:.2f}"
+        )
+        print("-----------------------------------------------------------------\n")
+
+    def setup(self, x0=None, niter=None, tol=1e-4, show=False):
+        """Setup solver
+
+        Parameters
+        ----------
+        x0 : :obj:`np.ndarray`, optional
+            Initial guess
+        niter : :obj:`int`, optional
+            Number of iterations
+        tol : :obj:`float`, optional
+            Tolerance on residual norm
+        show : :obj:`bool`, optional
+            Display setup log
+
+        """
+        self.tol = tol
+        self.niter = niter
+        if show:
+            self._print_setup()
+
+        # initialize solver
+        self.x0 = x0
+        if self.x0 is None:
+            self.x = self.ncp.zeros(self.Op.shape[1], dtype=self.y.dtype)
+            self.r = self.y.copy()
+        else:
+            self.x = self.x0.copy()
+            self.r = self.y - self.Op.matvec(self.x)
+        self.c = self.r.copy()
+        self.kold = self.ncp.abs(self.r.dot(self.r.conj()))
+
+        # create variables to track the residual norm and iterations
+        self.cost = []
+        self.cost.append(np.sqrt(self.kold))
+        self.iiter = 0
+
+    def step(self, show=False):
+        """Run one step of solver
+
+        Parameters
+        ----------
+        show : :obj:`bool`, optional
+            Display iteration log
+
+        """
+        Opc = self.Op.matvec(self.c)
+        cOpc = self.ncp.abs(self.c.dot(Opc.conj()))
+        a = self.kold / cOpc
+        self.x += a * self.c
+        self.r -= a * Opc
+        k = self.ncp.abs(self.r.dot(self.r.conj()))
+        b = k / self.kold
+        self.c = self.r + b * self.c
+        self.kold = k
+        self.iiter += 1
+        self.cost.append(np.sqrt(self.kold))
+        if show:
+            self._print_step()
+
+    def finalize(self, show=False):
+        """Finalize solver
+
+        Parameters
+        ----------
+        show : :obj:`bool`, optional
+            Display finalize log
+
+        """
+        if show:
+            self._print_finalize()
+        self.cost = np.array(self.cost)
+
+    def callback(self):
+        """Callback routine
+
+        This routine must be passed by the user as follows (when using the `solve`
+        method it will be automatically invoked after each step of the solve)
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from pylops.basicoperators import Identity
+        >>> from pylops.optimization.solver import CG
+        >>> def callback():

```

Comment:
What is the meaning of this callback? Why doesn't it take `x` input as before? It doesn't seem as useful to be as before.

Reply:
What if I want to use the callback for testing my solver against the ground truth, e.g.
```
history = []
def my_callback(x)
    history.append(np.linalg.norm(x - x_gt))
```
This is a good use-case, and perhaps you want to plot the error per iteration `plt.plot(history)`. You can't do this anymore, with this design.

Reply:
`Why doesn't it take x input as before?` Sorry this is a mistake, we should pass `x`

`What if I want to use the callback...?` This is the main use case of callback. You can simply use a lambda function 

```
def callback(xx, xhist):
    xhist.append(xx.copy())
...
cgsolve.callback = lambda x: callback(x, xhist)
```

This is how I would suggest to treat the callback in the current function based implementation and I don't think this can change for class based simply because inside the `solve` method we don't know what callback would look like and the best way is to enforce it to have one single input. Do you have a better suggestion?

Reply:
As long as `x` is passed there is no issue. I thought it was removed for some other reason. I don't want to change the interface.